dwave/optimization/model.pyi

```diff
@@ -12,29 +12,166 @@
 
 
 class Model:
-    def constant(self, array_like) -> Constant: ...
+    def __init__(self): ...
+
+    @property
+    def objective(self) -> ArraySymbol: ...
+    @property
+    def states(self) -> States: ...
+
+    def add_constraint(self, value: ArraySymbol): ...
+    def binary(self, shape: typing.Optional[_ShapeLike] = None) -> BinaryVariable: ...
+    def constant(self, array_like: numpy.typing.ArrayLike) -> Constant: ...
+    def decision_state_size(self) -> int: ...
+
+    def disjoint_bit_sets(
+        self, primary_set_size: int, num_disjoint_sets: int,
+        ) -> typing.Tuple[DisjointBitSets, typing.Tuple[DisjointBitSet, ...]]: ...
+
+    def disjoint_lists(
+        self, primary_set_size: int, num_disjoint_lists: int,
+        ) -> typing.Tuple[DisjointLists, typing.Tuple[DisjointList, ...]]: ...
+
+    @classmethod
+    def from_file(
+        cls,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        *,
+        check_header: bool = True,
+        ) -> Model: ...
+
+    def integer(
+        self,
+        shape: typing.Optional[_ShapeLike] = None,
+        lower_bound: typing.Optional[int] = None,
+        upper_bound: typing.Optional[int] = None,
+        ) -> IntegerVariable: ...
+
+    def into_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        *,
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ): ...
+
+    def is_locked(self) -> bool: ...
+    def iter_constraints(self) -> typing.Iterator[ArraySymbol]: ...
+    def iter_decisions(self) -> typing.Iterator[Symbol]: ...
+    def iter_symbols(self) -> typing.Iterator[Symbol]: ...
     def list(self, n: int) -> ListVariable: ...
-    def lock(self) -> None: ...
+    def lock(self) -> typing.ContextManager: ...
+    def minimize(self, value: ArraySymbol): ...
+    def num_constraints(self) -> int: ...
+    def num_decisions(self) -> int: ...
     def num_nodes(self) -> int: ...
-    def resize_states(self, n: int): ...
+    def num_symbols(self) -> int: ...
+
+    # dev note: this is underspecified, but it would be quite complex to fully
+    # specify the linear/quadratic so let's leave it alone for now.
+    def quadratic_model(self, x: ArraySymbol, quadratic, linear=None) -> QuadraticModel: ...
 
+    def set(
+        self,
+        n: int,
+        min_size: int = 0,
+        max_size: typing.Optional[int] = None,
+        ) -> SetVariable: ...
 
-class Constant(ArraySymbol):
-    ...
+    def state_size(self) -> int: ...
 
+    def to_file(
+        self,
+        *,
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ) -> typing.BinaryIO: ...
 
-class ListVariable(ArraySymbol):
-    def set_state(self, index: int, state: numpy.typing.ArrayLike) -> None: ...
+    # networkx might not be installed, so we just say we return an object.
+    def to_networkx(self) -> object: ...
 
+    def unlock(self): ...
```

Comment:
Is there a typing for an exit form contextmanager? (equivalent to `-> typing.ContextManager`)

Reply:
`unlock()` is called by the context manager, it doesn't return one. We do `lock(self) -> typing.ContextManager` which covers enter/exit.